### PR TITLE
Implemented Show fact source in the City facts page 

### PR DESCRIPTION
**Pull request by vyankatesh24**
# Description

Added one textview on facts screen and made some code chages related to issue.
If there are any UI change, please include the screenshots also.

Fixes #(issue) [#166](https://github.com/project-travel-mate/Travel-Mate/issues/166)

image

### CODE REVIEW

File: Android/app/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -126,6 +126,7 @@
 
     <!-- Sample Data -->
     <string name="sample_state">Delhi</string>
+    <string name="sample_source">service sutra</string>
```
Comment (Swati4star) [inline Android/app/src/main/res/values/strings.xml]: Remove this. I don't think this is required now.

File: Android/app/src/main/res/layout/fragment_funfact.xml [hunk 2]
```diff
@@ -51,4 +51,15 @@
...
         android:textSize="25sp" />
 
+    <TextView
+        android:id="@+id/source"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:fontFamily="sans-serif-light"
+        android:padding="10dp"
+        android:text="@string/sample_source"
+        android:textColor="#bbbbbb"
```
Comment (Swati4star) [inline Android/app/src/main/res/layout/fragment_funfact.xml]: add this color in colors.xml file & use that here

File: Android/app/src/main/res/layout/fragment_funfact.xml [hunk 3]
```diff
@@ -51,4 +51,15 @@
         android:textColor="#000000"
         android:textSize="25sp" />
 
+    <TextView
+        android:id="@+id/source"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:fontFamily="sans-serif-light"
+        android:padding="10dp"
+        android:text="@string/sample_source"
```
Comment (Swati4star) [inline Android/app/src/main/res/layout/fragment_funfact.xml]: Don't add any default value. Just keep it blank

File: Android/app/src/main/java/io/github/project_travel_mate/destinations/funfacts/FunfactFragment.java [hunk 4]
```diff
@@ -75,26 +75,37 @@ public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
         if (fact != null) {
             (holder.desc).setText(fact.getText());
             (holder.title).setText(fact.getTitle());
+            (holder.source).setText(fact.getSource());
```
Comment (Swati4star) [inline Android/app/src/main/java/io/github/project_travel_mate/destinations/funfacts/FunfactFragment.java]: Add a check before this line, 
if (fact.getSource() != null && fact.getSource() != "null")

File: Android/app/src/main/java/io/github/project_travel_mate/destinations/funfacts/FunfactFragment.java [hunk 5]
```diff
@@ -75,26 +75,37 @@ public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
...
+            (holder.source).setOnClickListener(view1 -> {
+
+                String url = fact.getSourceURL();
+                Uri sourceURI = Uri.parse(url);
+                Intent intent = new Intent(Intent.ACTION_VIEW, sourceURI);
+                if (intent.resolveActivity(Objects.requireNonNull(getActivity()).getPackageManager()) != null) {
+                    startActivity(intent);
+                }
```
Comment (Swati4star) [inline Android/app/src/main/java/io/github/project_travel_mate/destinations/funfacts/FunfactFragment.java]: if no intent is found, show a snackbar message, "No application to open this"

**Comment by Swati4star** [comment]
Please post a new screenshot also.

File: Android/app/src/main/res/layout/fragment_funfact.xml [hunk 6]
```diff
@@ -51,4 +51,31 @@
...
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:orientation="horizontal"
+        android:padding="10dp">
+
+        <TextView
+            android:id="@+id/text_source"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:text="@string/text_source"
```
Comment (Swati4star) [inline Android/app/src/main/res/layout/fragment_funfact.xml]: When there is no text source (it is null), the text view will show
**Source : **

This won't look good.

File: Android/app/src/main/java/objects/FunFact.java [hunk 7]
```diff
@@ -8,30 +8,41 @@
...
+
+
+    private String mSourceURL;
 
     /***
      * Initialize funfact object
      * @param title fact title
      * @param image fact image
      * @param text  face text
+     * @param source
```
Comment (Swati4star) [inline Android/app/src/main/java/objects/FunFact.java]: add sourceURL also here; along with description

File: Android/app/src/main/java/objects/FunFact.java [hunk 8]
```diff
@@ -8,30 +8,41 @@
     private String mTitle;
     private String mImage;
     private String mText;
+    private String mSource;
+
```
Comment (Swati4star) [inline Android/app/src/main/java/objects/FunFact.java]: Remove these 2 lines

File: Android/app/src/main/res/layout/fragment_funfact.xml (hunk 6 above)
Comment (Swati4star) [reply Android/app/src/main/res/layout/fragment_funfact.xml]: It's default value is gone. My bad. Ignore this comment.

File: Android/app/src/main/java/io/github/project_travel_mate/destinations/funfacts/FunfactFragment.java [hunk 9]
```diff
@@ -75,29 +76,48 @@ public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
         if (fact != null) {
             (holder.desc).setText(fact.getText());
             (holder.title).setText(fact.getTitle());
+            if (fact.getSource() != null && !fact.getSource().equals("null")) {
+                (holder.text_source).setVisibility(View.VISIBLE);
+                (holder.source).setText(fact.getSource());
+                (holder.source).setOnClickListener(view1 -> {
+
+                    String url = fact.getSourceURL();
```
Comment (Swati4star) [inline Android/app/src/main/java/io/github/project_travel_mate/destinations/funfacts/FunfactFragment.java]: move lines 84 to 93 in a new function (`openURL`), just like `shareImage` function.

**Comment by Swati4star** [comment]
@vyankatesh24 The checks are also failing. Look into those.

**Comment by vyankatesh24** [comment]
image

**Pull request merged**